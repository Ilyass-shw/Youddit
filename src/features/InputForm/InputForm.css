.input-form {
	display: flex;
	align-items: center;
}
input {
	background-color: #121212;
	color: #d1d1d1;
	border: 1px solid #303030;
	height: 1.75rem;
	font-size: 0.9rem;
	flex-basis: 34.5rem;
	border-radius: 2px 0 0 2px;
	padding: 2px 6px;
	min-width: 0;
}
.button-container {
	position: relative;
}
.search-button {
	background-color: #313131;
	border-radius: 0 2px 2px 0;
	height: 2.1rem;
	min-width: 4.06rem;
	border-inline: none;

	border-block: 1px solid #303030;
}
input:focus {
	border: 0.2px solid rgb(0, 89, 255);
	box-shadow: 0 0 10px #313131;
}
.search-label {
	padding: 0.4rem;
	color: #ffffff;
	background-color: #5a5a5ac9;
	border-radius: 2px;
	position: absolute;
	right: 7px;
	top: 38px;
	opacity: 0;
	transition: opacity 0.28s;
	font-size: 0.75rem;
	z-index: 2;
}
button:hover + p {
	opacity: 1;
}
