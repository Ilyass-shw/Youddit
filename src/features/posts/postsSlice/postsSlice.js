import { createAsyncThunk, createSlice } from "@reduxjs/toolkit";
import { client, getEndPoint, handlefetchedPosts } from "../../../api/api";

const initialState = {
	searchTerm: "memes",
	postList: [],
	filter: "relevance",
	firstFilter: "relevance",
	lastFilter: "comments",
	postFetchingStatus: "idle",
	error: null,
};

export const fetchPosts = createAsyncThunk("posts/fetchPosts", async (_, { getState }) => {
	const filter = getState().posts.filter;
	const searchTerm = getState().posts.searchTerm;

	const endPoint = getEndPoint(searchTerm, filter);
	const response = await client.get(endPoint);

	const fetchedPosts = handlefetchedPosts(response);
	return fetchedPosts;
});

const postsSlice = createSlice({
	name: "posts",
	initialState,
	reducers: {
		searchTermSet(state, action) {
			state.searchTerm = action.payload;
		},
		filterUpdated(state, action) {
			state.filter = action.payload;
		},
		firstFilterUpdated(state, action) {
			state.firstFilter = action.payload;
		},
		lastFilterUpdated(state, action) {
			state.lastFilter = action.payload;
		},
	},
	extraReducers: {
		[fetchPosts.pending]: (state, action) => {
			state.postFetchingStatus = "Loading";
		},
		[fetchPosts.fulfilled]: (state, action) => {
			state.postFetchingStatus = "succeeded";
			state.postList = action.payload;
		},
		[fetchPosts.rejected]: (state, action) => {
			state.postFetchingStatus = "failed";
			state.error = action.error.message;
		},
	},
});

export const { searchTermSet, filterUpdated, firstFilterUpdated, lastFilterUpdated } = postsSlice.actions;
export default postsSlice.reducer;

export const selectAllPosts = (state) => state.posts.postList;
