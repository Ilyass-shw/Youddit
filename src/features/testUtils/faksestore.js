import { configureStore } from "@reduxjs/toolkit";
import { createSlice } from "@reduxjs/toolkit";

const initialState = {
	firstFilter: "relevance",
	lastFilter: "comments",
	filter: "relevance",
};
const fakeSlice = createSlice({
	name: "posts",
	initialState,
	reducers: {},
});

export default configureStore({
	reducer: {
		posts: fakeSlice.reducer,
	},
});

export const makeCustomTestStore = (name, initialState, reducers = {}) => {
	const fakeSlice = createSlice({
		name,
		initialState,
		reducers,
	});

	return configureStore({
		reducer: {
			posts: fakeSlice.reducer,
		},
	});
};
