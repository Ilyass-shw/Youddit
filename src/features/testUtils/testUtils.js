import React from "react";
import { render as rtlRender } from "@testing-library/react";
import { Provider } from "react-redux";
import store, { makeCustomTestStore } from "./faksestore";

const TestProvider = ({ store, children }) => <Provider store={store}>{children}</Provider>;

export function render(ui, { store, ...otherOpts }) {
	return rtlRender(<TestProvider store={store}>{ui}</TestProvider>, otherOpts);
}

export function makeTestStore(name, initialState, reducers) {
	if (name && initialState) {
		const Store = makeCustomTestStore(name, initialState, reducers);
		const origDispatch = Store.dispatch;
		Store.dispatch = jest.fn(origDispatch);
		return Store;
	} else {
		const origDispatch = store.dispatch;
		store.dispatch = jest.fn(origDispatch);
		return store;
	}
}

// re-export everything
export * from "@testing-library/react";
