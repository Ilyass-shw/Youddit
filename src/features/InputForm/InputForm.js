import React, { useState } from "react";
import "./InputForm.css";
import { InlineIcon } from "@iconify/react";
import search16Filled from "@iconify-icons/fluent/search-16-filled";
import { fetchPosts, searchTermSet } from "../posts/postsSlice/postsSlice";
import { useDispatch } from "react-redux";
import { Transition } from "react-spring/renderprops";
export const InputForm = () => {
	const dispatch = useDispatch();

	const [term, setTerm] = useState("");
	const [mouseOnSearchButton, setMouseOnSearchButton] = useState(false);

	const handleChange = ({ target }) => {
		setTerm(target.value);
	};

	const handleOnClick = (e) => {
		if (term) {
			dispatch(searchTermSet(term));
			dispatch(fetchPosts());
		}
		e.preventDefault();
	};

	const show = true;
	return (
		<div className="input">
			<form action="#" className="input-form" autoComplete="off">
				<Transition
					config={{ duration: 500 }}
					items={show}
					from={{ flexBasis: "0rem" }}
					enter={{ flexBasis: "34.5rem" }}
				>
					{(show) =>
						show &&
						((props) => (
							<input
								style={props}
								type="text"
								name="search"
								value={term}
								onChange={handleChange}
								placeholder="Search"
							/>
						))
					}
				</Transition>

				<div className="button-container">
					<button
						data-testid="search-button"
						className="search-button"
						onMouseEnter={() => {
							setMouseOnSearchButton(true);
						}}
						onMouseLeave={() => {
							setMouseOnSearchButton(false);
						}}
						onClick={handleOnClick}
					>
						<InlineIcon
							icon={search16Filled}
							width="1.22rem"
							height="1.22rem"
							color={mouseOnSearchButton ? "#a3a3a3" : "#6f6f6f"}
						/>
					</button>
					<p className="search-label">Search</p>
				</div>
			</form>
		</div>
	);
};
