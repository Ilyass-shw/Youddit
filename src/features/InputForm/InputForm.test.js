import { jest } from "@jest/globals";
import { render, makeTestStore, screen } from "../testUtils/testUtils";
import React from "react";
import { InputForm } from "./InputForm.js";
import { searchTermSet } from "../posts/postsSlice/postsSlice";
import userEvent from "@testing-library/user-event";

jest.mock("@iconify-icons/fluent/search-16-filled", () => {
	return {
		default: "search16FilledMock",
	};
});



// jest.mock("../posts/postsSlice", () => {
// 	return {
// 		...jest.requireActual("../posts/postsSlice"),
// 		fetchPosts: () => "posts are fetched",
// 	};
// });

jest.mock("react-spring/renderprops", () => {
	return {
		Transition: ({ children, enter }) => {
			return <div data-testid="Transition">{children("InputForm")(enter)}</div>;
		},
	};
});

jest.mock("@iconify/react", () => {
	return { InlineIcon: () => <p>InlineIcon mock</p> };
});

let store;

describe("InputForm", () => {
	beforeEach(() => {
		store = makeTestStore();
	});

	it("should render without crashing", () => {
		render(<InputForm />, { store });
	});

	it("should match the snapshot", () => {
		const { container } = render(<InputForm />, { store });

		expect(container).toMatchSnapshot();
	});

	it("should have an empty initial value for input ", () => {
		render(<InputForm />, { store });
		const input = screen.getByPlaceholderText("Search");
		expect(input.value).toBe("");
	});

	it("should not dispatch anything when clicked without defined term", () => {
		render(<InputForm />, { store });
		const input = screen.getByPlaceholderText("Search");
		expect(input.value).toBe("");
		const submitButton = screen.getByTestId("search-button");
		expect(store.dispatch).toBeCalledTimes(0);

		userEvent.click(submitButton);
		expect(store.dispatch).toBeCalledTimes(0);
	});

	it("should dispatch and fetch posts according to the typed term after clicking the search button", () => {
		render(<InputForm />, { store });
		const input = screen.getByPlaceholderText("Search");

		expect(input).toHaveValue("");

		userEvent.type(input, "memes");

		const submitButton = screen.getByTestId("search-button");

		expect(store.dispatch).toBeCalledTimes(0);

		userEvent.click(submitButton);

		expect(store.dispatch).toHaveBeenNthCalledWith(1, searchTermSet("memes"));
		expect(store.dispatch).toBeCalledTimes(2);
	});
});
