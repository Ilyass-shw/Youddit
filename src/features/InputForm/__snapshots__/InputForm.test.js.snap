// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputForm should match the snapshot 1`] = `
<div>
  <div
    class="input"
  >
    <form
      action="#"
      autocomplete="off"
      class="input-form"
    >
      <div
        data-testid="Transition"
      >
        <input
          name="search"
          placeholder="Search"
          style="flex-basis: 34.5rem;"
          type="text"
          value=""
        />
      </div>
      <div
        class="button-container"
      >
        <button
          class="search-button"
          data-testid="search-button"
        >
          <p>
            InlineIcon mock
          </p>
        </button>
        <p
          class="search-label"
        >
          Search
        </p>
      </div>
    </form>
  </div>
</div>
`;
